### PR TITLE
Add generic object API

### PR DESCRIPTION
**Pull request by r-rojo**
This adds an API for getting generic objects from a token or handle.
One of the main advantages of using the fpgaObject API is that objects
can be buffered. New data can be read from the underlying driver if the
FPGA_OBJECT_SYNC flag is used.
Additionally, it can keep file descriptors of open sysfs files exported
by the OPAE kernel driver.
Finally, the string key used to locate an object can have wildcards but
can only be used if a user of the API explicitly asks for globbing with
the FPGA_OBJECT_GLOB flag.

These are the following functions added in this change:
```C
fpga_result fpgaTokenGetObject(fpga_token token, const char *name,
		               fpga_object *object, int flags);
```
Used to create an object from a key. The key can be a sysfs path
relative to the token sysfs path. This will be opened with read-only
access.

```C
fpga_result fpgaHandleGetObject(fpga_handle handle, const char *name,
		                fpga_object *object, int flags);
```
Used to create an object from a key. The key can be a sysfs path
relative to the handle sysfs path. This will be opened with read-write
access.

```C
fpga_result fpgaObjectGetObject(fpga_object parent, fpga_handle handle,
				const char *name, fpga_object *object,
				int flags);
```
Used to create an object from a key. The key can be a sysfs path
relative to the parent sysfs path. This will be opened with read-write
access if a valid handle is passed in, and read-only if a null handle is
passed in.

```C
fpga_result fpgaDestroyObject(fpga_object *obj);
```
Used to free memory created by one of the object creating functions
above.

```C
fpga_result fpgaObjectRead(fpga_object obj, uint8_t *buffer, size_t offset,
			   size_t len, int flags);
```
Used to read data from the object. This is only valid for leaf objects
(not containers).

```C
fpga_result fpgaObjectRead64(fpga_object obj, uint64_t *value, int flags);
```
Convenience function to read a 64-bit integer. Note, the data may be
parsed from a string if the flag FPGA_OBJECT_TEXT is used.

```C
fpga_result fpgaObjectWrite64(fpga_object obj, uint64_t value, int flags);
```
Convenience function to write a 64-bit integer. Note, the data may be
converted to a string if the flag FPGA_OBJECT_TEXT is used.

### CODE REVIEW

File: common/include/opae/sysobject.h [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+/**
+ * @file sysobject.h
+ * @brief Functions to read/write from system objects.
+ * On Linux systems with the OPAE kernel driver, this is used to access sysfs
+ * nodes created by the driver.
+ */
+#ifndef __SYSOBJECT_H__
```
Comment (tswhison) [inline common/include/opae/sysobject.h]: A nit: can we use __FPGA_SYSOBJECT_H__ to further isolate the namespace?

File: common/include/opae/sysobject.h [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+ * parameter is invalid. FPGA_NOT_FOUND if an object cannot be found with the
+ * given key. FPGA_NOT_SUPPORTED if this function is not supported by the
+ * current implementation of this API.
+ *
+ */
+fpga_result fpgaHandleGetObject(fpga_handle handle, const char *name,
+				fpga_object *object, int flags);
+
+/**
+ * @brief Create an `fpga_obect` data structure from a parent object.
```
Comment (tswhison) [inline common/include/opae/sysobject.h]: s/fpga_obect/fpga_object/

File: common/include/opae/sysobject.h [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+ * @param[out] object Pointer to memory to store the object in
+ * @param[in] flags Control behavior of object identification and creation
+ * FPGA_OBJECT_GLOB is used to indicate that the name should be treated as a
+ * globbing expression.  FPGA_OBJECT_RECURSE_ONE indicates that subobjects be
+ * created for objects one level down from the object identified by name.
+ * FPGA_OBJECT_RECURSE_ALL indicates that subobjects be created for all objects
+ * below the current object identified by name.
+ *
+ * @return FPGA_OK on success. FPGA_INVALID_PARAM if any of the supplied
+ * parameter is invalid. FPGA_NOT_FOUND if an object cannot be found with the
```
Comment (tswhison) [inline common/include/opae/sysobject.h]: s/parameter/parameters/

File: common/include/opae/sysobject.h [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+ * @brief Create an `fpga_obect` data structure from a parent object.
+ * An `fpga_object` is a handle to an FPGA resource which can be an attribute
+ * or a register. If the handle parameter is NULL then the object is created
+ * with read-only access. If the handle is a valid handle, then it will be
+ * created with read-write access.
+ *
+ * @param[in] parent A parent container `fpga_object`.
+ * @param[in] handle Handle identifying a resource (accelerator or device).
+ * @param[in] name A key identifying a sub-object of the parent container.
+ * @param[out] objecta Pointer to memory to store the object in.
```
Comment (tswhison) [inline common/include/opae/sysobject.h]: s/objecta/object a/

File: common/include/opae/sysobject.h [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+ * parameters is invalid
+ */
+fpga_result fpgaObjectRead64(fpga_object obj, uint64_t *value, int flags);
+
+/**
+ * @brief Write 64-bit value to an FPGA object
+ *
+ * @param[in] obj An fpga_object instance.
+ * @param[in] value The value to write to the object
+ * @param[in] flags Flags that control how the object is read
```
Comment (tswhison) [inline common/include/opae/sysobject.h]: s/read/written/

File: common/include/opae/types.h [hunk 6]
```diff
@@ -165,6 +165,8 @@ typedef void *fpga_event_handle;
  * an accelerator resource. The error API provides functions to retrieve these
  * information structures from a particular resource.
  */
+
+typedef void *fpga_object;
```
Comment (tswhison) [inline common/include/opae/types.h]: This is inserted between fpga_error_info and its Doxygen comment.  Doxygen won't do the right thing in this case.  Move it above the comment or below the fpga_error_info definition.

File: libopae/src/sysfs.c [hunk 7]
```diff
@@ -29,13 +29,17 @@
 #endif // HAVE_CONFIG_H
```
Comment (tswhison) [inline libopae/src/sysfs.c]: Looks like you ran a formatter on this file.  I can't tell which changes are required for your object API.

File: libopae/src/sysobject.c [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+
+#include "common_int.h"
+#include "log_int.h"
+#include "sysfs_int.h"
+#include "types_int.h"
+#include "safe_string/safe_string.h"
+#include <opae/types_enum.h>
+#include <opae/sysobject.h>
+
+#define SYSOBJ_ASSERT_NOT_NULL(var)                                            \
```
Comment (tswhison) [inline libopae/src/sysobject.c]: We already have ASSERT_NOT_NULL macro, which has a return statement.  Can we use that instead?

File: libopae/src/sysobject.c [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+	close(fd);
+
+	return res;
+}
+
+
+fpga_result fpgaTokenGetObject(fpga_token token, const char *name,
+			       fpga_object *object, int flags)
+{
+	(void)flags;
```
Comment (tswhison) [inline libopae/src/sysobject.c]: Can we use UNUSED_PARAM for consistency?

File: libopae/src/sysobject.c [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+	}
+
+	return make_sysfs_object(objpath, name, object, flags, NULL);
+}
+
+
+fpga_result fpgaHandleGetObject(fpga_token handle, const char *name,
+				fpga_object *object, int flags)
+{
+	(void)flags;
```
Comment (tswhison) [inline libopae/src/sysobject.c]: Can we use UNUSED_PARAM for consistency?

File: libopae/src/sysobject.c [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+	}
+
+	return make_sysfs_object(objpath, name, object, flags, handle);
+}
+
+fpga_result fpgaObjectGetObject(fpga_object parent, fpga_handle handle,
+				const char *name, fpga_object *object,
+				int flags)
+{
+	(void)flags;
```
Comment (tswhison) [inline libopae/src/sysobject.c]: Can we use UNUSED_PARAM for consistency?

File: libopae/src/sysobject.c [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+	if (res) {
+		return res;
+	}
+
+	res = cat_sysfs_path(objpath, name);
+	if (res) {
+		return res;
+	}
+
+	// for (i = 0; i < _obj->size; ++i) {
```
Comment (tswhison) [inline libopae/src/sysobject.c]: remove dead code

File: libopae/src/types_int.h [hunk 13]
```diff
@@ -210,5 +211,20 @@ struct token_map {
 	struct token_map *next;
 };
 
+typedef enum { FPGA_SYSFS_DIR = 0, FPGA_SYSFS_FILE } fpga_sysfs_type;
+
+struct _fpga_object {
+	pthread_mutex_t lock;
+  fpga_handle handle;
```
Comment (tswhison) [inline libopae/src/types_int.h]: spacing

File: samples/object_api.c [hunk 14]
```diff
@@ -0,0 +1,397 @@
...
+	fpga_object object;
+	uint64_t value;
+	uint64_t delta;
+} named_object;
+
+typedef struct {
+	const char *name;
+	fpga_token token;
+	fpga_object object;
+        uint8_t bus;
```
Comment (tswhison) [inline samples/object_api.c]: spacing

File: samples/object_api.c [hunk 15]
```diff
@@ -0,0 +1,397 @@
...
+        uint8_t function;
+	named_object objects[MAX_GROUP_OBJECTS];
+	size_t count;
+} metric_group;
+
+typedef struct {
+	fpga_token token;
+	metric_group *groups;
+	size_t count;
+        fpga_object clock;
```
Comment (tswhison) [inline samples/object_api.c]: spacing

File: samples/object_api.c [hunk 16]
```diff
@@ -0,0 +1,397 @@
...
+} token_group;
+
+// ├── cache
```
Comment (tswhison) [inline samples/object_api.c]: specific to DCP?  remove

File: samples/object_api.c [hunk 17]
```diff
@@ -0,0 +1,397 @@
...
+
+metric_group *init_metric_group(fpga_token token, const char *name, metric_group *group)
+{
+	fpga_properties props;
+	fpga_result res;
+
+	group->token = token;
+	group->name = name;
+	group->count = 0;
+        if (fpgaGetProperties(token, &props) == FPGA_OK) {
```
Comment (tswhison) [inline samples/object_api.c]: spacing

File: samples/object_api.c [hunk 18]
```diff
@@ -0,0 +1,397 @@
...
+		goto out;
+	}
+	res = fpgaObjectGetObject(group->object, handle, "freeze", &object, 0);
+	if (res != FPGA_OK) {
+		goto out_close;
+	}
+	res = fpgaObjectWrite64(object, value, FPGA_OBJECT_TEXT);
+	if ((other_res = fpgaDestroyObject(&object)) != FPGA_OK) {
+		print_err("error destorying object", other_res);
+	};
```
Comment (tswhison) [inline samples/object_api.c]: s/;//

File: samples/object_api.c [hunk 19]
```diff
@@ -0,0 +1,397 @@
...
+	// create a new filter of type ACCELERATOR
+	res = fpgaGetProperties(NULL, &filter);
+	ON_ERR_GOTO(res, out, "Creating fpga_properties");
+	ADD_TO_CLEANUP(fpgaDestroyProperties, &filter);
+
+	// set filter to accelerator object
+	res = fpgaPropertiesSetObjectType(filter, FPGA_DEVICE);
+	ON_ERR_GOTO(res, out_free, "Setting ObjectType");
+
+	if (options.bus > 0) {
```
Comment (tswhison) [inline samples/object_api.c]: bus 0 is valid.

File: samples/object_api.c [hunk 20]
```diff
@@ -0,0 +1,397 @@
...
+	}
+	int count = 100;
+	while (--count > 0) {
+		for (i = 0; i < num_matches; ++i) {
+			for (j = 0; j < metrics[i].count; ++j) {
+				freeze_group(&metrics[i].groups[j], 1);
+				print_counters(metrics[i].clock, &metrics[i].groups[j]);
+				freeze_group(&metrics[i].groups[j], 0);
+			}
+			usleep(1E6);
```
Comment (tswhison) [inline samples/object_api.c]: s/1E6/1000000/

File: common/include/opae/sysobject.h (hunk 1 above)
Comment (r-rojo) [reply common/include/opae/sysobject.h]: 👍

File: common/include/opae/sysobject.h (hunk 2 above)
Comment (r-rojo) [reply common/include/opae/sysobject.h]: 👍

File: common/include/opae/sysobject.h (hunk 3 above)
Comment (r-rojo) [reply common/include/opae/sysobject.h]: 👍

File: common/include/opae/sysobject.h (hunk 4 above)
Comment (r-rojo) [reply common/include/opae/sysobject.h]: 👍

File: common/include/opae/sysobject.h (hunk 5 above)
Comment (r-rojo) [reply common/include/opae/sysobject.h]: 👍

File: samples/object_api.c (hunk 16 above)
Comment (r-rojo) [reply samples/object_api.c]: 👍

File: common/include/opae/types.h (hunk 6 above)
Comment (r-rojo) [reply common/include/opae/types.h]: 👍

File: libopae/src/sysfs.c (hunk 7 above)
Comment (r-rojo) [reply libopae/src/sysfs.c]: 😢

File: libopae/src/sysobject.c (hunk 8 above)
Comment (r-rojo) [reply libopae/src/sysobject.c]: 👍

File: libopae/src/sysobject.c (hunk 9 above)
Comment (r-rojo) [reply libopae/src/sysobject.c]: 👍

File: libopae/src/sysobject.c (hunk 10 above)
Comment (r-rojo) [reply libopae/src/sysobject.c]: 👍

File: common/include/opae/sysobject.h [hunk 21]
```diff
@@ -40,23 +41,79 @@ extern "C" {
 
 /**
  * @brief Create an `fpga_object` data structures. An `fpga_object`
- * is a handle to an FPGA resource which can be an attribute or register.
- * This object is read-only.
+ * is a handle to an FPGA resource which can be an attribute or register or
+ * drvier attribute. This object is read-only.
```
Comment (tswhison) [inline common/include/opae/sysobject.h]: s/drvier/driver/

File: common/include/opae/types.h [hunk 22]
```diff
@@ -171,6 +171,21 @@ struct fpga_error_info {
 	bool can_clear;                   /** whether error can be cleared */
 };
 
+/** Object pertaining to an FPGA resource as identified by a unique name
+ *
+ * An `fpga_object` represents either a device attribute or a container of
+ * attributes. Similar to filesystems, a '/' may be used to seperate objects in
+ * an object hierarchy. Once on object is aquired, it may be used to read or
```
Comment (tswhison) [inline common/include/opae/types.h]: s/aquired/acquired/

File: common/include/opae/types.h [hunk 23]
```diff
@@ -171,6 +171,21 @@ struct fpga_error_info {
...
+ * write data in a resource attribute or to query sub-objects if the object is
+ * a container object. The data in an object is buffered and will be kept
+ * around until the object is destroyed. Additonaly, the data in an attribute
```
Comment (tswhison) [inline common/include/opae/types.h]: s/Additonaly/Additionally/

File: common/include/opae/types.h [hunk 24]
```diff
@@ -171,6 +171,21 @@ struct fpga_error_info {
...
+ * can by synchronized from the owning resource using the FPGA_OBJECT_SYNC flag
+ * during read operations.  The name identifying the object is unique with
+ * repsect to the resource that owns it. A parent resource may be identified by
```
Comment (tswhison) [inline common/include/opae/types.h]: s/repsect/respect/

File: libopae/plugins/xfpga/sysfs.c [hunk 25]
```diff
@@ -778,26 +780,65 @@ struct _fpga_object *alloc_fpga_object(const char *sysfspath, const char *name)
 {
 	struct _fpga_object *obj = calloc(1, sizeof(struct _fpga_object));
```
Comment (tswhison) [inline libopae/plugins/xfpga/sysfs.c]: Just a nit - if you're init'ing all of the struct members below, then you can use malloc instead (skips zeroing of members).  On the other hand, if you add new struct members in the future, then they wouldn't be initialized to 0.

File: libopae/plugins/xfpga/sysobject.c [hunk 26]
```diff
@@ -48,28 +48,131 @@
...
+
+	fd = open(objpath, O_RDONLY);
+	if (fd < 0) {
+		return FPGA_NOT_FOUND;
+	}
+
+	count = eintr_read(fd, buffer, *len);
+	if (count < (ssize_t)*len) {
+		if (count < 0) {
+			FPGA_ERR("Error with pread operation: %s",
```
Comment (tswhison) [inline libopae/plugins/xfpga/sysobject.c]: s/pread/read/ ? Or, are you still using partial reads?

File: samples/CMakeLists.txt [hunk 27]
```diff
@@ -71,14 +71,17 @@ endif()
 
 add_executable(hello_fpga hello_fpga.c)
 if(BUILD_LIBOPAE_C)
-  target_link_libraries(hello_fpga ${libjson-c_LIBRARIES} uuid ${CMAKE_THREAD_LIBS_INIT} opae-c)
+  target_link_libraries(hello_fpga ${lib${libjson-c_LIBRARIES}_LIBRARIES} uuid ${CMAKE_THREAD_LIBS_INIT} opae-c)
```
Comment (tswhison) [inline samples/CMakeLists.txt]: yikes!

File: samples/CMakeLists.txt [hunk 28]
```diff
@@ -71,14 +71,17 @@ endif()
...
 elseif(BUILD_ASE)
-  target_link_libraries(hello_fpga ${libjson-c_LIBRARIES} uuid ${CMAKE_THREAD_LIBS_INIT} opae-c-ase)
+  target_link_libraries(hello_fpga ${lib${libjson-c_LIBRARIES}_LIBRARIES} uuid ${CMAKE_THREAD_LIBS_INIT} opae-c-ase)
```
Comment (tswhison) [inline samples/CMakeLists.txt]: yikes!

File: samples/CMakeLists.txt [hunk 29]
```diff
@@ -71,17 +71,20 @@ endif()
 
 add_executable(hello_fpga hello_fpga.c)
 if(BUILD_LIBOPAE_C)
-  target_link_libraries(hello_fpga ${libjson-c_LIBRARIES} uuid ${CMAKE_THREAD_LIBS_INIT} opae-c)
+  target_link_libraries(hello_fpga ${libson-c_LIBRARIES} uuid ${CMAKE_THREAD_LIBS_INIT} opae-c)
```
Comment (tswhison) [inline samples/CMakeLists.txt]: s/libson/libjson/

File: testing/CMakeLists.txt [hunk 30]
```diff
@@ -43,7 +43,7 @@ add_library(xfpga-static STATIC
               ${OPAE_SDK_SOURCE}/libopae/plugins/xfpga/reconf.c
               ${OPAE_SDK_SOURCE}/libopae/plugins/xfpga/reset.c
               ${OPAE_SDK_SOURCE}/libopae/plugins/xfpga/sysfs.c
-              #${OPAE_SDK_SOURCE}/libopaeplugins/xfpga//sysobject.c
+              ${OPAE_SDK_SOURCE}/libopae/plugins/xfpga//sysobject.c
```
Comment (tswhison) [inline testing/CMakeLists.txt]: s,xfpga//,xfpga/,

File: common/include/opae/types_enum.h [hunk 31]
```diff
@@ -144,14 +144,12 @@ enum fpga_reconf_flags {
...
- * Object Synchronization flags
- *
- * These flags can be passed to the fpgaObjectRead() function.
- */
 enum fpga_object_read_flags {
-	/** Synchronize the object contents by reading from the underlying object. */
-	FPGA_OBJECT_SYNC = (1u << 0)
+  FPGA_OBJECT_SYNC = (1u << 0), /**< Synchronize data from driver */
+  FPGA_OBJECT_GLOB = (1u << 1), /**< Treat names as glob expressions */
+  FPGA_OBJECT_TEXT = (1u << 2), /**< Parse or convert numeric data as text */
```
Comment (lzhan55) [inline common/include/opae/types_enum.h]: when this flag is set, should we use the (*fpgaObjectRead)(...) to parse the numeric data as text?

File: samples/object_api.c [hunk 32]
```diff
@@ -0,0 +1,338 @@
...
+	metric_group *groups;
+	size_t count;
+	fpga_object clock;
+} token_group;
+
+struct {
+	bool freeze;
+	int bus;
+	float interval_sec;
+} options = { false, -1, 1.0 };
```
Comment (lzhan55) [inline samples/object_api.c]: What is this options about? When freeze is set, what the bus and interval_sec values mean?

Comment (r-rojo) [reply samples/object_api.c]: Freeze is used to tell the sample to freeze (take a snapshot) of the counters before reading them.
Writing to the "freeze" sysfs file tells the driver to snapshot all counters in the sysfs directory.

bus is used to filter when enumerating. interval_usec is the microseconds to sleep in between reading perf counters.

File: common/include/opae/types_enum.h (hunk 31 above)
Comment (r-rojo) [reply common/include/opae/types_enum.h]: This flag is only checked in the fpgaObjectRead64 and fpgaObjectWrite64 functions. fpgaObjectRead is used to read raw bytes without parsing it as ascii text.

**Pull request merged**